sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DisposeFromDispose.cs

```diff
@@ -104,11 +104,231 @@ public void CloseResource()
+
+        public void CleanUp()
+        {
+            this.fs.Dispose(); // Compliant - the Dispose method is not accessible
+        }
+
+        private void Dispose()
+        {
+        }
+    }
+
+    public struct DisposableStruct : IDisposable
+    {
+        private FileStream fs;
+        public void OpenResource(string path)
+        {
+            this.fs = new FileStream(path, FileMode.Open);
+        }
+        public void CloseResource()
+        {
+            this.fs.Close();
+        }
+
+        public void CleanUp()
+        {
+            this.fs.Dispose(); // Noncompliant
+        }
+
+        public void Dispose()
+        {
+            // method added to satisfy demands of interface
+        }
+    }
+
+    public struct DisposableStructCorrect : IDisposable
+    {
+        private FileStream fs;
+        public void OpenResource(string path)
+        {
+            this.fs = new FileStream(path, FileMode.Open);
+        }
+        public void CloseResource()
+        {
+            this.fs.Close();
+        }
+
+        public void Dispose()
+        {
+            this.fs.Dispose(); // compliant
+        }
+    }
+
+    public ref struct MyTestDisposableRefStruct
+    {
+        public void Dispose()
+        {
+        }
+    }
+
+    public ref struct RefStructResourceHolder
+    {
+        // ref structs can only be fields in other ref structs
+
+        private MyTestDisposableRefStruct foo;
+
+        public void OpenResource(string path)
+        {
+            this.foo = new MyTestDisposableRefStruct();
+        }
+
+        public void CleanUp()
+        {
+            this.foo.Dispose(); // Noncompliant
+        }
+
+        public void Dispose()
+        {
+            // this method makes this struct a disposable ref struct in C# 8
+        }
+    }
+
+    public ref struct NotDisposableRefStruct
+    {
+        public void Dispose(bool shouldDispose)
+        {
+        }
+        public string Dispose()
+        {
+            return "";
+        }
+    }
+
+    public ref struct NotDisposableRefStructHolder1
+    {
+        private NotDisposableRefStruct foo;
+
+        public void OpenResource(string path)
+        {
+            this.foo = new NotDisposableRefStruct();
+        }
+
+        public void CleanUp()
+        {
+            this.foo.Dispose(true); // Ok
+        }
+        public void Dispose()
+        {
+        }
+    }
+
+    public ref struct NotDisposableRefStructHolder2
+    {
+        private NotDisposableRefStruct foo;
+
+        public void OpenResource(string path)
+        {
+            this.foo = new NotDisposableRefStruct();
+        }
+
+        public void CleanUp()
+        {
+            var x = foo.Dispose(); // Ok
+        }
+        public void Dispose()
+        {
+        }
+    }
+
+
+
+    public ref struct AnotherDisposableRefStruct
+    {
+        public void Dispose() { }
+        public void Dispose(bool x) { }
+        public void Dispose(string y) { }
```

Comment:
I didn't add tests for `public void Dispose<T>() { }` (with different type parameter arity) , but I think that's fine, I wouldn't go through another iteration